src/instructions/register.rs

```diff
@@ -26,31 +21,49 @@ pub trait Register:
     fn min_value() -> Self;
     fn max_value() -> Self;
 
-    // By default, PartialOrd/Ord compares unsigned value, which is the same
-    // as shl/shr.
-    fn signed_cmp(&self, other: &Self) -> Ordering;
+    // Conditional operations, if the condition evaluated here is true, R::one()
+    // will be emitted, otherwise R::zero() will be emitted
+    fn eq(self, other: Self) -> Self;
+    fn less_than(self, other: Self) -> Self;
+    fn less_than_signed(self, other: Self) -> Self;
+    fn logical_not(self) -> Self;
+
+    // self here should be the result of one of the conditional operations, if
+    // self is R::one(), true_value will be returned, otherwise false_value will
+    // be returned. No values other than R::one() or R::zero() should be accepted
+    // as self here.
+    fn cond(self, true_value: Self, false_value: Self) -> Self;
+
+    fn overflowing_add(self, rhs: Self) -> Self;
+    fn overflowing_sub(self, rhs: Self) -> Self;
+    fn overflowing_mul(self, rhs: Self) -> Self;
+
+    // Those 4 methods should implement RISC-V's overflowing strategies:
+    // +---------------------------------------------------------------------------------------+
+    // | Condition              | Dividend  | Divisor | DIVU[W] | REMU[W] |  DIV[W]   | REM[W] |
+    // +------------------------+-----------+---------+---------+---------+-----------+--------+
+    // | Division by zero       |     x     |    0    | 2**L-1  |    x    |    -1     |   x    |
+    // +------------------------+-----------+---------+---------+---------+-----------+--------+
+    // | Overflow (signed only) | −2**(L−1) |   −1    |    -    |    -    | -2**(L-1) |   0    |
+    // +---------------------------------------------------------------------------------------+
+    fn overflowing_div(self, rhs: Self) -> Self;
+    fn overflowing_rem(self, rhs: Self) -> Self;
+    fn overflowing_div_signed(self, rhs: Self) -> Self;
+    fn overflowing_rem_signed(self, rhs: Self) -> Self;
```

Comment:
Why no overflow indicator returned?

Reply:
We want all the returned information here to be contained within Register type, so we can provide enough information for JIT(if we have returned `overflow` indicator here, and use it in a Rust conditional in `instructions` package, JIT might not have enough information to transform code). And since those div/rem instructions are the only place where we actually use overflow indicators, we can change the implementation to put div/rem logic within registers, and expose no overflow information.